core/src/read.jl

```diff
@@ -1036,20 +1036,40 @@ function Allocation(db::DB, config::Config, graph::MetaGraph)::Allocation
         optimization_type = String[],
     )
 
-    mean_flows = Dict{Tuple{NodeID, NodeID}, Base.RefValue{Float64}}()
+    mean_input_flows = Dict{Tuple{NodeID, NodeID}, Base.RefValue{Float64}}()
 
     # Find edges which serve as sources in allocation
     for edge_metadata in values(graph.edge_data)
         (; subnetwork_id_source, edge) = edge_metadata
         if subnetwork_id_source != 0
-            mean_flows[edge] = Ref(0.0)
+            mean_input_flows[edge] = Ref(0.0)
         end
     end
 
     # Find basins with a level demand
     for node_id in values(graph.vertex_labels)
         if has_external_demand(graph, node_id, :level_demand)[1]
-            mean_flows[(node_id, node_id)] = Ref(0.0)
+            mean_input_flows[(node_id, node_id)] = Ref(0.0)
+        end
+    end
+
+    mean_realized_flows = Dict{Tuple{NodeID, NodeID}, Base.RefValue{Float64}}()
+
+    # Find edges that realize a demand
+    for edge_metadata in values(graph.edge_data)
+        (; type, edge) = edge_metadata
+
+        src_id, dst_id = edge
+        user_demand_inflow = (type == EdgeType.flow) && (dst_id.type == NodeType.UserDemand)
+        level_demand_inflow =
+            (type == EdgeType.control) && (src_id.type == NodeType.LevelDemand)
+        flow_demand_inflow =
+            (type == EdgeType.flow) && has_external_demand(graph, dst_id, :flow_demand)[1]
+
+        if user_demand_inflow || flow_demand_inflow
+            mean_realized_flows[edge] = Ref(0.0)
+        elseif level_demand_inflow
+            mean_realized_flows[(dst_id, dst_id)] = Ref(0.0)
         end
```

Comment:
Can you explain why you went for Ref here instead of a Float? Is it for the ergonomics of being able to do

```julia
value = dict[key]
value[] += x
value[] /= dt
```

```julia
value = dict[key]
value += x
value /= dt
dict[key] = value
```

i.e. having to put it back in? I worry a bit about the performance of this approach.

Reply:
I got rid of the use of `RefValue{Float64}`, but now there are more dictionary lookups, e.g.

https://github.com/Deltares/Ribasim/blob/8e337aa1c00b80ca1cbc3be52117bc3f903dacaf/core/src/callback.jl#L487-L500

Is there a way to circumvent this?

Reply:
Dict lookups are quite fast, so I doubt it's a serious issue. I haven't looked into it, but perhaps the compiler can figure out it can optimize this with a single lookup. Also I just tried this, and surprisingly it works. Could be of course that there are still two lookups behind this syntax:

```
julia> dict = Dict{Int, Int}(1 => 1)
Dict{Int64, Int64} with 1 entry:
  1 => 1

julia> dict[1] += 1
2

julia> dict
Dict{Int64, Int64} with 1 entry:
  1 => 2
```